File: HoTT-Model/src/main/java/de/treichels/hott/model/helicopter/HelicopterModel.java
/**
 * HoTT Transmitter Config Copyright (C) 2013 Oliver Treichel
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package de.treichels.hott.model.helicopter;

import de.treichels.hott.model.BaseModel;
import de.treichels.hott.model.Switch;
import de.treichels.hott.model.enums.ModelType;
import de.treichels.hott.model.enums.PitchMin;
import de.treichels.hott.model.enums.RotorDirection;
import de.treichels.hott.model.enums.SwashplateType;

import javax.xml.bind.annotation.XmlElementWrapper;
import javax.xml.bind.annotation.XmlIDREF;
import javax.xml.bind.annotation.XmlRootElement;
import java.util.Arrays;

/**
 * @author Oliver Treichel &lt;oli@treichels.de&gt;
 */
@XmlRootElement
public class HelicopterModel extends BaseModel {
    private static final long serialVersionUID = 1L;

    private int autorotationC1TriggerPosition;
    private Switch autorotationSwitch;
    private int autorotationTailPosition;
    private int autorotationThrottlePosition;
    private int expoThrottleLimit;
    private Switch markerSwitch;
    private PitchMin pitchMin;
    private HelicopterProfiTrim[] profiTrim;
    private RotorDirection rotorDirection;
    private boolean swashplateLinearization;
    private SwashplateMix swashplateMix;
    private SwashplateType swashplateType;
    private int throttleLimitWarning;
    private boolean throttleMarkerActive;
    private int throttleMarkerPosition;

    public HelicopterModel() {
        this(ModelType.Helicopter);
    }

    public HelicopterModel(final ModelType cmodelType) {
        super(cmodelType);
    }

    @Override
    public boolean equals(final Object obj) {
        if (this == obj) return true;
        if (!super.equals(obj)) return false;
        if (getClass() != obj.getClass()) return false;
        final HelicopterModel other = (HelicopterModel) obj;
        if (autorotationC1TriggerPosition != other.autorotationC1TriggerPosition) return false;
        if (autorotationSwitch == null) {
            if (other.autorotationSwitch != null) return false;
        } else if (!autorotationSwitch.equals(other.autorotationSwitch)) return false;
        if (autorotationTailPosition != other.autorotationTailPosition) return false;
        if (autorotationThrottlePosition != other.autorotationThrottlePosition) return false;
        if (expoThrottleLimit != other.expoThrottleLimit) return false;
        if (markerSwitch == null) {
            if (other.markerSwitch != null) return false;
        } else if (!markerSwitch.equals(other.markerSwitch)) return false;
        if (pitchMin != other.pitchMin) return false;
        if (!Arrays.equals(profiTrim, other.profiTrim)) return false;
        if (rotorDirection != other.rotorDirection) return false;
        if (swashplateLinearization != other.swashplateLinearization) return false;
        if (swashplateMix == null) {
            if (other.swashplateMix != null) return false;
        } else if (!swashplateMix.equals(other.swashplateMix)) return false;
        if (swashplateType != other.swashplateType) return false;
        return throttleLimitWarning == other.throttleLimitWarning;
    }

    public int getAutorotationC1TriggerPosition() {
        return autorotationC1TriggerPosition;
    }

    @XmlIDREF
    public Switch getAutorotationSwitch() {
        return autorotationSwitch;
    }

    public int getAutorotationTailPosition() {
        return autorotationTailPosition;
    }

    public int getAutorotationThrottlePosition() {
        return autorotationThrottlePosition;
    }

    public int getExpoThrottleLimit() {
        return expoThrottleLimit;
    }

    @XmlIDREF
    public Switch getMarkerSwitch() {
        return markerSwitch;
    }

    public PitchMin getPitchMin() {
        return pitchMin;
    }

    @XmlElementWrapper(name = "profitrims")
    public HelicopterProfiTrim[] getProfiTrim() {
        return profiTrim;
    }

    public RotorDirection getRotorDirection() {
        return rotorDirection;
    }

    public SwashplateMix getSwashplateMix() {
        return swashplateMix;
    }

    public SwashplateType getSwashplateType() {
        return swashplateType;
    }

    public int getThrottleLimitWarning() {
        return throttleLimitWarning;
    }

    public int getThrottleMarkerPosition() {
        return throttleMarkerPosition;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = super.hashCode();
        result = prime * result + autorotationC1TriggerPosition;
        result = prime * result + (autorotationSwitch == null ? 0 : autorotationSwitch.hashCode());
        result = prime * result + autorotationTailPosition;
        result = prime * result + autorotationThrottlePosition;
        result = prime * result + expoThrottleLimit;
        result = prime * result + (markerSwitch == null ? 0 : markerSwitch.hashCode());
        result = prime * result + (pitchMin == null ? 0 : pitchMin.hashCode());
        result = prime * result + Arrays.hashCode(profiTrim);
        result = prime * result + (rotorDirection == null ? 0 : rotorDirection.hashCode());
        result = prime * result + (swashplateLinearization ? 1231 : 1237);
        result = prime * result + (swashplateMix == null ? 0 : swashplateMix.hashCode());
        result = prime * result + (swashplateType == null ? 0 : swashplateType.hashCode());
        result = prime * result + throttleLimitWarning;
        return result;
    }

    public boolean isSwashplateLinearization() {
        return swashplateLinearization;
    }

    public boolean isThrottleMarkerActive() {
        return throttleMarkerActive;
    }

    public void setAutorotationC1TriggerPosition(final int autorotationPosition) {
        autorotationC1TriggerPosition = autorotationPosition;
    }

    public void setAutorotationSwitch(final Switch autorotatonSwitch) {
        autorotationSwitch = autorotatonSwitch;
    }

    public void setAutorotationTailPosition(final int autorotationTailPosition) {
        this.autorotationTailPosition = autorotationTailPosition;
    }

    public void setAutorotationThrottlePosition(final int autorotationThrottlePosition) {
        this.autorotationThrottlePosition = autorotationThrottlePosition;
    }

    public void setExpoThrottleLimit(final int expoThrottleLimit) {
        this.expoThrottleLimit = expoThrottleLimit;
    }

    public void setMarkerSwitch(final Switch markerSwitch) {
        this.markerSwitch = markerSwitch;
    }

    public void setPitchMin(final PitchMin pitchMin) {
        this.pitchMin = pitchMin;
    }

    public void setProfiTrim(final HelicopterProfiTrim[] profiTrim) {
        this.profiTrim = profiTrim;
    }

    public void setRotorDirection(final RotorDirection rotorDirection) {
        this.rotorDirection = rotorDirection;
    }

    public void setSwashplateLinearization(final boolean swashplateLinearization) {
        this.swashplateLinearization = swashplateLinearization;
    }

    public void setSwashplateMix(final SwashplateMix swashplateMix) {
        this.swashplateMix = swashplateMix;
    }

    public void setSwashplateType(final SwashplateType swashplateType) {
        this.swashplateType = swashplateType;
    }

    public void setThrottleLimitWarning(final int throttleLimitWarning) {
        this.throttleLimitWarning = throttleLimitWarning;
    }

    public void setThrottleMarkerActive(final boolean throttleMarkerActive) {
        this.throttleMarkerActive = throttleMarkerActive;
    }

    public void setThrottleMarkerPosition(final int throttleMarkerPosition) {
        this.throttleMarkerPosition = throttleMarkerPosition;
    }
}

File: HoTT-Model/src/main/java/de/treichels/hott/model/helicopter/HelicopterModel.kt
/**
 * HoTT Transmitter Config Copyright (C) 2013 Oliver Treichel
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program. If not, see <http:></http:>//www.gnu.org/licenses/>.
 */
package de.treichels.hott.model.helicopter

import de.treichels.hott.model.BaseModel
import de.treichels.hott.model.Switch
import de.treichels.hott.model.enums.ModelType
import de.treichels.hott.model.enums.PitchMin
import de.treichels.hott.model.enums.RotorDirection
import de.treichels.hott.model.enums.SwashplateType
import javax.xml.bind.annotation.XmlElementWrapper
import javax.xml.bind.annotation.XmlIDREF
import javax.xml.bind.annotation.XmlRootElement

/**
 * @author Oliver Treichel &lt;oli@treichels.de&gt;
 */
@XmlRootElement
open class HelicopterModel(
        modelType: ModelType = ModelType.Helicopter,
        var autorotationC1TriggerPosition: Int = 0,
        @get:XmlIDREF
        var autorotationSwitch: Switch = Switch(),
        var autorotationTailPosition: Int = 0,
        var autorotationThrottlePosition: Int = 0,
        var expoThrottleLimit: Int = 0,
        @get:XmlIDREF
        var markerSwitch: Switch = Switch(),
        var pitchMin: PitchMin? = null,
        @get:XmlElementWrapper(name = "profitrims")
        var profiTrim: List<HelicopterProfiTrim> = emptyList(),
        var rotorDirection: RotorDirection? = null,
        var isSwashplateLinearization: Boolean = false,
        var swashplateMix: SwashplateMix? = null,
        var swashplateType: SwashplateType? = null,
        var throttleLimitWarning: Int = 0,
        var isThrottleMarkerActive: Boolean = false,
        var throttleMarkerPosition: Int = 0
) : BaseModel(modelType) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is HelicopterModel) return false
        if (!super.equals(other)) return false

        if (autorotationC1TriggerPosition != other.autorotationC1TriggerPosition) return false
        if (autorotationSwitch != other.autorotationSwitch) return false
        if (autorotationTailPosition != other.autorotationTailPosition) return false
        if (autorotationThrottlePosition != other.autorotationThrottlePosition) return false
        if (expoThrottleLimit != other.expoThrottleLimit) return false
        if (markerSwitch != other.markerSwitch) return false
        if (pitchMin != other.pitchMin) return false
        if (profiTrim != other.profiTrim) return false
        if (rotorDirection != other.rotorDirection) return false
        if (isSwashplateLinearization != other.isSwashplateLinearization) return false
        if (swashplateMix != other.swashplateMix) return false
        if (swashplateType != other.swashplateType) return false
        if (throttleLimitWarning != other.throttleLimitWarning) return false
        if (isThrottleMarkerActive != other.isThrottleMarkerActive) return false
        if (throttleMarkerPosition != other.throttleMarkerPosition) return false

        return true
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + autorotationC1TriggerPosition
        result = 31 * result + autorotationSwitch.hashCode()
        result = 31 * result + autorotationTailPosition
        result = 31 * result + autorotationThrottlePosition
        result = 31 * result + expoThrottleLimit
        result = 31 * result + markerSwitch.hashCode()
        result = 31 * result + (pitchMin?.hashCode() ?: 0)
        result = 31 * result + profiTrim.hashCode()
        result = 31 * result + (rotorDirection?.hashCode() ?: 0)
        result = 31 * result + isSwashplateLinearization.hashCode()
        result = 31 * result + (swashplateMix?.hashCode() ?: 0)
        result = 31 * result + (swashplateType?.hashCode() ?: 0)
        result = 31 * result + throttleLimitWarning
        result = 31 * result + isThrottleMarkerActive.hashCode()
        result = 31 * result + throttleMarkerPosition
        return result
    }
}
